### DISCUSSION TITLE
What is SPI_HALF_DUPLEX in Zephyr

### DISCUSSION THREAD

**Discussion by RichardSWheatley** [zephyrproject-rtos/zephyr#69985] [Polls]
Zephyr says:

```
    /**
    * @name SPI duplex mode
    * @{
    *
    * Some controllers support half duplex transfer, which results in 3-wire usage.
    * By default, full duplex will prevail.
    */
    #define SPI_FULL_DUPLEX		(0U << 11)
    #define SPI_HALF_DUPLEX		(1U << 11)
    /** @} */
```

**Comment by ThreeEights** [comment]
Well, in the SPI world, "3-wire" means there are 3 wires between the host and the peripheral:  CS, CLK, and a single DATA line, instead of the more common 4 wires with separate DATA_OUT and DATA_IN lines.  A 4-wire configuration could run in either full-duplex or half-duplex, but from what I've seen, most Zephyr SPI drivers don't support half-duplex operation.  A 3-wire configuration, of course, can only run in half-duplex;  in my case, trying to talk to a 3-wire device with a Raspberry Pi Pico, I have flip the direction of the GPIO line between sending a command and receiving the data (and making sure the flip it back before sending the next command!).

**Reply by RichardSWheatley** [reply]
I agree but it doesn't say 3-wire configuration. It says "3-wire usage".

I think they need to fix the wording.

But you didn't answer the poll.

**Reply by klesh** [reply]
@ThreeEights Hi, I am struggling to get the half-duplex to work on RP2040. Would you mind sharing the details of how to do it?
I can get it to work using QMK, but not luck using ZMK with Zephyr3.5.

**Reply by ThreeEights** [reply]
@klesh - I'm guessing you are talking about the ZMK keyboard firmware, https://zmk.dev/  If so ... I'm not familiar with that.  The Zephyr SPI drivers for the RP2040 assume that they're operating as the bus master.  Are you trying to connect a keyboard to the RP2040 using an SPI bus?

**Reply by klesh** [reply]
Yes, I was referring to the ZMK keyboard firmware, it is based Zephyr 3.5, I was not looking to connect split keyboards using SPI, but an optical sensor.

I figured it might be a driver problem, because nrf52840 worked fine while rp2040 didn't. After some digging, I saw this: https://github.com/zmkfirmware/zephyr/blob/0fa4cc26d8b1e5243570ccc0f7f66c4c428d2961/drivers/spi/spi_pl022.c#L358

So, I might have to give up on ZMK on RP2040 for the time being 😂 .

**Comment by tobiwan88** [comment]
THe original definition of spi is only full duplex 4 wire (good old motorola spi) https://onlinedocs.microchip.com/pr/GUID-835917AF-E521-4046-AD59-DCB458EB8466-en-US-1/index.html?GUID-E4682943-46B9-4A20-A62C-33E8FD3343A3

From the missuses implementations i saw they always define 3 wire as a half duplex, E.g. https://www.st.com/en/microcontrollers-microprocessors/stm32l0-series/documentation.html 
<img width="757" alt="Screenshot 2024-03-09 at 08 19 22" src="https://github.com/zephyrproject-rtos/zephyr/assets/6998679/c960adc8-4c0b-4293-8970-8732cd4dbb6c">
